File: task_1_my_threadpool/java_version/MyThreadpool.java
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

public class MyThreadpool {
    private final MyThread[] mWorkers;
    private final BlockingQueue<Runnable> mQueue = new LinkedBlockingQueue<>();

    public MyThreadpool(int threadsNumber) {
        if (threadsNumber > 0) {
            mWorkers = new MyThread[threadsNumber];
        } else {
            System.out.println("Invalid value for threadsNumber argument: " + threadsNumber);
            mWorkers = null;
            return;
        }

        for (int i = 0; i < threadsNumber; i++) {
            mWorkers[i] = new MyThread();
            mWorkers[i].start();
        }
    }

    public void execute(Runnable task) {
        synchronized (mQueue){
            mQueue.add(task);
            mQueue.notify();
        }
    }

    public void terminate(){
        for (MyThread worker:mWorkers) {
            worker.interrupt();
        }
    }

    private class MyThread extends Thread {
        public void run(){
            try{
                while (true){
                    Runnable task;

                    synchronized (mQueue){
                        if (mQueue.isEmpty()) {
                            mQueue.wait();
                        }
                        task = mQueue.poll();
                    }
                    if (task != null)
                        task.run();
                }
            } catch (InterruptedException e) {
                return;
            }
        }
    }
}

File: task_1_my_threadpool/kotlin_version/MyThreadPool.kt
import java.util.concurrent.LinkedBlockingQueue

class MyThreadPool(threadsNumber: Int) {
    val mQueue: LinkedBlockingQueue<Runnable> = LinkedBlockingQueue<Runnable>()
    private val mLock = Object()
    var mWorkers: Array<MyThread>

    init{
        mWorkers = emptyArray()

        for (x in 0 until threadsNumber) {
            mWorkers += MyThread()
            mWorkers[x].start()
        }
    }

    fun execute(task: Runnable) {
        synchronized(mLock){
            mQueue.add(task)
            mLock.notify()
        }
    }

    inner class MyThread:Thread() {
        override fun run(){
            try {
                while (!interrupted()) {
                    var task: Runnable

                    synchronized(mLock) {
                        if (mQueue.isEmpty()) {
                            mLock.wait()
                        }
                        task = mQueue.poll()
                    }
                    task.run()
                }
            } catch (e: InterruptedException) {
                return
            }
        }
    }

    fun terminate() {
        mWorkers.forEach {it.interrupt()}
    }
}